### PR TITLE
Add XACML DAO Interfaces

### PR DESCRIPTION
**Pull request by Niru-D**
### Proposed changes in this pull request

- Add DAO interfaces to support multiple implementations .

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java [hunk 1]
```diff
@@ -44,7 +44,7 @@
 /**
  * Entitlement PDP related admin services are exposed
  */
-public class EntitlementAdminService {
+public class  EntitlementAdminService {
```
Comment (rksk) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java:47]:
```suggestion
public class EntitlementAdminService {
```

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java [hunk 2]
```diff
@@ -24,6 +25,7 @@
 import org.wso2.carbon.context.CarbonContext;
 import org.wso2.carbon.context.RegistryType;
 import org.wso2.carbon.identity.entitlement.common.EntitlementConstants;
+import org.wso2.carbon.identity.entitlement.dao.*;
```
Comment (rksk) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java:28]: Let's avoid wildcard imports

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/PolicyDAO.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.List;
+
+
+/**
+ * This interface supports the management of XACML policies.
+ */
+public interface PolicyDAO extends PolicyFinderModule {
+
+
+    /**
```
Comment (Yoshani) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/PolicyDAO.java:35]: Use the following format for doc comments
```
/**
     * Check whether a resource exists in the given path in the tenant registry.
     *
     * @param path         Path of the resource.
     * @param tenantDomain Tenant domain of the resource.
     * @return True if the resource exists, false otherwise.
     * @throws IdentityRuntimeException If an error occurs while checking the existence of the resource.
     */
```

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java [hunk 4]
```diff
@@ -426,7 +426,7 @@ public PDPDataHolder getPDPData() {
      * @param finder
      * @return
      */
-    public PolicyFinderDataHolder getPolicyFinderData(String finder) {
+    public PolicyFinderDataHolder getPolicyFinderData(String finder) throws EntitlementException {
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java:429]: Can we avoid updating this method signature

Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java:429]: Better if we can avoid this signature change.

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java [hunk 5]
```diff
@@ -527,8 +527,8 @@ public PIPFinderDataHolder getPIPResourceFinderData(String finder) {
      */
     public String getGlobalPolicyAlgorithm() throws EntitlementException {
 
-        return EntitlementAdminEngine.getInstance().
```
[the remark this answers is not in the document]
Comment (darshanasbg) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java:530]: This change dont needed with https://github.com/wso2/carbon-identity-framework/pull/5681/commits/517da62e5e80ca7376373e275f4e6797af1af331#diff-406fe136e4a3478e44cd1f6c4b1ecfaa230c904a350077d69b5e07b7861d1c54

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java [hunk 6]
```diff
@@ -539,7 +539,7 @@ public String getGlobalPolicyAlgorithm() throws EntitlementException {
      */
     public void setGlobalPolicyAlgorithm(String policyCombiningAlgorithm) throws EntitlementException {
 
-        EntitlementAdminEngine.getInstance().
-                getPolicyDataStore().setGlobalPolicyAlgorithm(policyCombiningAlgorithm);
+        ConfigDAO config = new RegistryConfigDAOImpl();
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java:542]: This one also wont be needed with https://github.com/wso2/carbon-identity-framework/pull/5681/commits/517da62e5e80ca7376373e275f4e6797af1af331#diff-406fe136e4a3478e44cd1f6c4b1ecfaa230c904a350077d69b5e07b7861d1c54

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementNotificationExtension.java [hunk 7]
```diff
@@ -36,7 +37,7 @@
  * registered in entitlement.properties file.
  */
 @SuppressWarnings("unused")
-public class EntitlementNotificationExtension implements PAPStatusDataHandler {
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementNotificationExtension.java:39]: This change wont be needed as we discussed to keep the `PAPStatusDataHandler` untouched.

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java [hunk 8]
```diff
@@ -285,11 +278,11 @@ public PolicyDTO getPolicy(String policyId, boolean isPDPPolicy) throws Entitlem
      */
     public PolicyDTO getPolicyByVersion(String policyId, String version) throws EntitlementException {
 
-        PolicyDTO policyDTO = null;
+        PolicyDTO policyDTO;
 
         try {
-            PolicyVersionManager versionManager = EntitlementAdminEngine.getInstance().getVersionManager();
-            policyDTO = versionManager.getPolicy(policyId, version);
+            PolicyDAO policyStore = new RegistryPolicyDAOImpl();
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java:284]: Should handle similar to https://github.com/wso2/carbon-identity-framework/pull/5673/files#r1594986499

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java [hunk 9]
```diff
@@ -369,21 +361,18 @@ public void removePolicy(String policyId, boolean dePromote) throws EntitlementE
         }
         handleStatus(EntitlementConstants.StatusTypes.DELETE_POLICY, oldPolicy, true, null);
 
-        //remove versions
-        EntitlementAdminEngine.getInstance().getVersionManager().deletePolicy(policyId);
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java:373]: Is this change intentional? @Niru-D

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java [hunk 10]
```diff
@@ -425,10 +414,12 @@ public String[] getAllPolicyIds(String searchString) throws EntitlementException
      */
     public PublisherDataHolder getSubscriber(String subscribeId) throws EntitlementException {
 
-        PolicyPublisher publisher = EntitlementAdminEngine.getInstance().getPolicyPublisher();
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java:428]: Should handle similar to https://github.com/wso2/carbon-identity-framework/pull/5673/files#r1594986499

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java [hunk 11]
```diff
@@ -437,59 +428,64 @@ public PublisherDataHolder getSubscriber(String subscribeId) throws EntitlementE
      * @throws EntitlementException throws, if fails
      */
     public String[] getSubscriberIds(String searchString) throws EntitlementException {
-        PolicyPublisher publisher = EntitlementAdminEngine.getInstance().getPolicyPublisher();
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java:440]: Should handle similar to https://github.com/wso2/carbon-identity-framework/pull/5673/files#r1594986499

Comment (darshanasbg) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java:440]: check the rest as well

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dto/PublisherDataHolder.java [hunk 12]
```diff
@@ -24,6 +24,8 @@
 import org.wso2.carbon.core.util.CryptoUtil;
 import org.wso2.carbon.registry.core.Resource;
 
+import java.sql.ResultSet;
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dto/PublisherDataHolder.java:27]: Remove these unused imports.. @Niru-D

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dto/PolicyStoreDTO.java [hunk 13]
```diff
@@ -37,6 +37,10 @@ public class PolicyStoreDTO {
 
     private boolean setActive;
 
+    private String version;
+
+    private String policyType;
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dto/PolicyStoreDTO.java:42]: Do we need this anymore. There are no usages for the policy type. @Niru-D

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java [hunk 14]
```diff
@@ -527,8 +527,8 @@ public PIPFinderDataHolder getPIPResourceFinderData(String finder) {
      */
     public String getGlobalPolicyAlgorithm() throws EntitlementException {
 
-        return EntitlementAdminEngine.getInstance().
-                getPolicyDataStore().getGlobalPolicyAlgorithmName();
+        ConfigDAO config = new RegistryConfigDAOImpl();
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementAdminService.java:530]: Initializing this should be handled more centrally as mentioned in https://github.com/wso2/carbon-identity-framework/pull/5673/files#r1595074332

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/policy/finder/CarbonPolicyFinder.java [hunk 15]
```diff
@@ -143,17 +146,8 @@ protected boolean removeEldestEntry(Map.Entry eldest) {
...
-            PolicyDataStore policyDataStore;
-            Map<PolicyDataStore, Properties> dataStoreModules = EntitlementServiceComponent.
-                    getEntitlementConfig().getPolicyDataStore();
-            if (dataStoreModules != null && dataStoreModules.size() > 0) {
-                policyDataStore = dataStoreModules.entrySet().iterator().next().getKey();
-            } else {
-                policyDataStore = new DefaultPolicyDataStore();
-            }
-            policyCombiningAlgorithm = policyDataStore.getGlobalPolicyAlgorithm();
+            ConfigDAO config = new RegistryConfigDAOImpl();
```
[the remark this answers is not in the document]
Comment (darshanasbg) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/policy/finder/CarbonPolicyFinder.java:149]: Check about loosing the pluggablity aspect as well.

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java (hunk 9 above)
Comment (Niru-D) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementPolicyAdminService.java:373]: Yes, removing policy versions is done inside the policy removal method implementation of the new interface.

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dto/PolicyStoreDTO.java (hunk 13 above)
Comment (Niru-D) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dto/PolicyStoreDTO.java:42]: Removed.

**Comment by Yoshani** [comment]
Closing since these changes are incorporated in other PRs. Tracked in https://github.com/wso2/product-is/issues/20000
